File: app/src/test/java/com/dpis/module/BatchScopeRequestCoordinatorTest.java
package com.dpis.module;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import io.github.libxposed.service.XposedService;

public class BatchScopeRequestCoordinatorTest {
    @Test
    public void modernRequestsMissingPackagesInOneBatch() {
        RecordingHost host = new RecordingHost();
        FakeRequester requester = new FakeRequester();
        requester.scope = List.of("com.example.in.scope");
        BatchScopeRequestCoordinator coordinator =
                new BatchScopeRequestCoordinator(host, requester, true);

        BatchScopeRequestCoordinator.Result result = coordinator.requestMissingScope(
                List.of("com.example.in.scope", "com.example.missing.one", "com.example.missing.two"));

        assertTrue(result.requestStarted);
        assertEquals(List.of("com.example.missing.one", "com.example.missing.two"),
                requester.requestedPackages);
        assertEquals(1, requester.requestCallCount);
        assertEquals(R.string.quick_template_scope_request_started, host.toastIds.get(0).intValue());
    }

    @Test
    public void compatDoesNotRequestAndShowsManualGuidance() {
        RecordingHost host = new RecordingHost();
        FakeRequester requester = new FakeRequester();
        BatchScopeRequestCoordinator coordinator =
                new BatchScopeRequestCoordinator(host, requester, false);

        BatchScopeRequestCoordinator.Result result = coordinator.requestMissingScope(
                List.of("com.example.app"));

        assertFalse(result.requestStarted);
        assertTrue(result.manualRequired);
        assertEquals(0, requester.requestCallCount);
        assertEquals(R.string.quick_template_scope_manual_required, host.toastIds.get(0).intValue());
    }

    @Test
    public void unknownScopeDoesNotRollbackAndShowsManualGuidance() {
        RecordingHost host = new RecordingHost();
        FakeRequester requester = new FakeRequester();
        requester.throwOnGetScope = true;
        BatchScopeRequestCoordinator coordinator =
                new BatchScopeRequestCoordinator(host, requester, true);

        BatchScopeRequestCoordinator.Result result = coordinator.requestMissingScope(
                List.of("com.example.app"));

        assertFalse(result.requestStarted);
        assertTrue(result.manualRequired);
        assertEquals(0, requester.requestCallCount);
        assertEquals(R.string.quick_template_scope_manual_required, host.toastIds.get(0).intValue());
    }

    @Test
    public void partialApprovalRefreshesActualScopeState() {
        RecordingHost host = new RecordingHost();
        FakeRequester requester = new FakeRequester();
        requester.scope = List.of();
        BatchScopeRequestCoordinator coordinator =
                new BatchScopeRequestCoordinator(host, requester, true);

        coordinator.requestMissingScope(List.of("com.example.one", "com.example.two"));
        requester.listener.onScopeRequestApproved(List.of("com.example.one"));

        assertEquals(1, host.requestAppsLoadCount);
        assertEquals(R.string.quick_template_scope_request_approved,
                host.toastIds.get(host.toastIds.size() - 1).intValue());
    }

    private static final class FakeRequester implements BatchScopeRequestCoordinator.ScopeRequester {
        List<String> scope = List.of();
        List<String> requestedPackages = List.of();
        int requestCallCount;
        boolean throwOnGetScope;
        XposedService.OnScopeEventListener listener;

        @Override
        public List<String> getScope() {
            if (throwOnGetScope) {
                throw new RuntimeException("scope unavailable");
            }
            return scope;
        }

        @Override
        public void requestScope(List<String> packages, XposedService.OnScopeEventListener listener) {
            requestCallCount++;
            requestedPackages = new ArrayList<>(packages);
            this.listener = listener;
        }
    }

    private static final class RecordingHost implements BatchScopeRequestCoordinator.Host {
        final ArrayList<Integer> toastIds = new ArrayList<>();
        int requestAppsLoadCount;

        @Override
        public void showToast(int messageResId, Object... formatArgs) {
            toastIds.add(messageResId);
        }

        @Override
        public void requestAppsLoad() {
            requestAppsLoadCount++;
        }

        @Override
        public void runOnUiThread(Runnable runnable) {
            runnable.run();
        }
    }
}

File: app/src/test/java/com/dpis/module/templates/BatchScopeRequestCoordinatorTest.kt
package com.dpis.module.templates

import com.dpis.module.R
import io.github.libxposed.service.XposedService
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class BatchScopeRequestCoordinatorTest {
    @Test
    fun modernRequestsMissingPackagesInOneBatch() {
        val host = RecordingHost()
        val requester = FakeRequester(scopePackages = listOf("com.example.in.scope"))
        val coordinator = BatchScopeRequestCoordinator(host, requester, true)

        val result = coordinator.requestMissingScope(
            listOf("com.example.in.scope", "com.example.missing.one", "com.example.missing.two"),
        )

        assertTrue(result.requestStarted)
        assertEquals(listOf("com.example.missing.one", "com.example.missing.two"), requester.requestedPackages)
        assertEquals(1, requester.requestCallCount)
        assertEquals(R.string.quick_template_scope_request_started, host.toastIds.first())
        requester.listener!!.onScopeRequestFailed("dismissed")
    }

    @Test
    fun compatDoesNotRequestAndShowsManualGuidance() {
        val host = RecordingHost()
        val requester = FakeRequester()
        val result = BatchScopeRequestCoordinator(host, requester, false)
            .requestMissingScope(listOf("com.example.app"))

        assertFalse(result.requestStarted)
        assertTrue(result.manualRequired)
        assertEquals(0, requester.requestCallCount)
        assertEquals(R.string.quick_template_scope_manual_required, host.toastIds.first())
    }

    @Test
    fun unknownScopeDoesNotRollbackAndShowsManualGuidance() {
        val host = RecordingHost()
        val requester = FakeRequester(throwOnGetScope = true)
        val result = BatchScopeRequestCoordinator(host, requester, true)
            .requestMissingScope(listOf("com.example.app"))

        assertFalse(result.requestStarted)
        assertTrue(result.manualRequired)
        assertEquals(0, requester.requestCallCount)
        assertEquals(R.string.quick_template_scope_manual_required, host.toastIds.first())
    }

    @Test
    fun partialApprovalRefreshesActualScopeState() {
        val host = RecordingHost()
        val requester = FakeRequester()
        val coordinator = BatchScopeRequestCoordinator(host, requester, true)

        coordinator.requestMissingScope(listOf("com.example.one", "com.example.two"))
        requester.listener!!.onScopeRequestApproved(mutableListOf("com.example.one"))

        assertEquals(1, host.requestAppsLoadCount)
        assertEquals(R.string.quick_template_scope_request_approved, host.toastIds.last())
    }

    @Test
    fun ignoresBatchWhileAnotherScopeRequestIsAwaitingConfirmation() {
        val host = RecordingHost()
        val requester = FakeRequester()
        val coordinator = BatchScopeRequestCoordinator(host, requester, true)

        coordinator.requestMissingScope(listOf("com.example.one"))
        val second = coordinator.requestMissingScope(listOf("com.example.two"))

        assertFalse(second.requestStarted)
        assertFalse(second.manualRequired)
        assertEquals(1, requester.requestCallCount)
        assertEquals(R.string.scope_request_pending, host.toastIds.last())
        requester.listener!!.onScopeRequestFailed("dismissed")
    }

    private class FakeRequester(
        var scopePackages: List<String> = emptyList(),
        var throwOnGetScope: Boolean = false,
    ) : BatchScopeRequestCoordinator.ScopeRequester {
        var requestedPackages: List<String> = emptyList()
        var requestCallCount = 0
        var listener: XposedService.OnScopeEventListener? = null

        override fun getScope(): List<String> {
            check(!throwOnGetScope) { "scope unavailable" }
            return scopePackages
        }

        override fun requestScope(
            packages: List<String>,
            listener: XposedService.OnScopeEventListener,
        ) {
            requestCallCount++
            requestedPackages = packages.toList()
            this.listener = listener
        }
    }

    private class RecordingHost : BatchScopeRequestCoordinator.Host {
        val toastIds = mutableListOf<Int>()
        var requestAppsLoadCount = 0

        override fun showToast(messageResId: Int, vararg formatArgs: Any?) {
            toastIds += messageResId
        }

        override fun requestAppsLoad() {
            requestAppsLoadCount++
        }

        override fun runOnUiThread(runnable: Runnable) = runnable.run()
    }
}
